File: dart/2015/day01b/day01b.dart
#!/usr/bin/env dart

import 'dart:io';

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

int process(String filename) {
    int floors = 0;
    int pos = 0;
    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        for ( final ch in line.runes ) {
            pos++;
            if ( ch == '('.codeUnitAt(0) ) {
                floors++;
            } else if ( ch == ')'.codeUnitAt(0) ) {
                floors--;
            }
            if ( floors < 0 ) {
                return pos;
            }
        }
    }
    return 0;
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}

File: r/2015/day01b/day01b.r
#!/usr/bin/env Rscript

usage <- function() {
    args <- commandArgs()
    progname <- sub("^--file=", "", args[grep("^--file=", args)])
    cat("usage:", progname, "<input file>\n")
    quit(status = 1)
}

process <- function(filename) {
    floors <- 0
    pos <- 0
    lines <- suppressWarnings(readLines(filename))
    for ( line in lines ) {
        chars <- strsplit(line, "")[[1]]
        for ( ch in chars ) {
            pos <- pos + 1
            if ( ch == '(' ) {
                floors <- floors + 1
            } else if ( ch == ')' ) {
                floors <- floors - 1
            }
            if ( floors < 0 ) {
                return(pos);
            }
        }
    }
    return(0)
}

main <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    if ( length(args) < 1 ) {
        usage()
    }
    filename <- args[1]
    result <- process(filename)
    cat("result =", result, "\n")
}

main()
